src/mem_table.rs

```diff
@@ -216,17 +221,30 @@ impl KVTable {
 
     /// Inserts a value, returning as soon as the value is written to the memtable but before
     /// it is flushed to durable storage.
-    fn put_or_delete(&self, key: Bytes, value: ValueDeletable, attrs: RowAttributes) {
+    fn put_or_delete(
+        &self,
+        key: Bytes,
+        value: ValueDeletable,
+        attrs: RowAttributes,
+    ) -> Result<(), SlateDBError> {
         let key_len = key.len();
         self.size.fetch_add(
             key_len + value.len() + sizeof_attributes(&attrs),
             Ordering::Relaxed,
         );
 
-        // it is safe to use fetch_max here to update the last tick
-        // because the monotonicity is enforced when generating the clock tick
-        // (see [crate::utils::MonotonicClock::now])
-        attrs.ts.map(|tick| self.last_tick.fetch_max(tick, SeqCst));
+        if let Some(row_tick) = attrs.ts {
+            let max_tick = self.last_tick.fetch_max(row_tick, SeqCst);
+            // we should never try to insert out of order data because the
+            // monotonicity is enforced when generating the clock tick
+            // (see [crate::utils::MonotonicClock::now])
+            if max_tick > row_tick {
+                return Err(InvalidClockTick {
```

Comment:
I'd feel better if this was an invalid db state error - it really indicates some sort of bug in slatedb right? `InvalidClockTick` is more an indication of a problem with the configured clock